src/librbd/api/Migration.cc

```diff
@@ -0,0 +1,1434 @@
+  if (r == -ENOENT) {
+    r = 0;
+  }
+  if (r < 0) {
+    lderr(m_cct) << "failed removing migration header: " << cpp_strerror(r)
+                 << dendl;
+    return r;
+  }
+
+  image_ctx->notify_update();
+
+  return 0;
+}
+
+template <typename I>
+int Migration<I>::unlink_src_image() {
+  if (m_src_old_format) {
+    return v1_unlink_src_image();
+  } else {
+    return v2_unlink_src_image();
+  }
+}
+
+template <typename I>
+int Migration<I>::v1_unlink_src_image() {
+  ldout(m_cct, 10) << dendl;
+
+  int r = tmap_rm(m_src_io_ctx, m_src_image_name);
+  if (r < 0) {
+    lderr(m_cct) << "failed removing " << m_src_image_name << " from tmap: "
+                 << cpp_strerror(r) << dendl;
+    return r;
+  }
+
+  return 0;
+}
+
+template <typename I>
+int Migration<I>::v2_unlink_src_image() {
+  ldout(m_cct, 10) << dendl;
+
+  m_src_image_ctx->owner_lock.get_read();
+  if (m_src_image_ctx->exclusive_lock != nullptr &&
+      m_src_image_ctx->exclusive_lock->is_lock_owner()) {
+    C_SaferCond ctx;
+    m_src_image_ctx->exclusive_lock->release_lock(&ctx);
+    m_src_image_ctx->owner_lock.put_read();
+    int r = ctx.wait();
+     if (r < 0) {
+      lderr(m_cct) << "error releasing exclusive lock: " << cpp_strerror(r)
+                   << dendl;
+      return r;
+     }
+  } else {
+    m_src_image_ctx->owner_lock.put_read();
+  }
+
+  int r = trash_move(m_src_io_ctx, RBD_TRASH_IMAGE_SOURCE_MIGRATION,
+                     m_src_image_ctx->name, 0);
+  if (r < 0) {
+    lderr(m_cct) << "failed moving image to trash: " << cpp_strerror(r)
+                 << dendl;
+    return r;
+  }
+
+  return 0;
+}
+
+template <typename I>
+int Migration<I>::relink_src_image() {
+  if (m_src_old_format) {
+    return v1_relink_src_image();
+  } else {
+    return v2_relink_src_image();
+  }
+}
+
+template <typename I>
+int Migration<I>::v1_relink_src_image() {
+  ldout(m_cct, 10) << dendl;
+
+  int r = tmap_set(m_src_io_ctx, m_src_image_name);
+  if (r < 0) {
+    lderr(m_cct) << "failed adding " << m_src_image_name << " to tmap: "
+                 << cpp_strerror(r) << dendl;
+    return r;
+  }
+
+  return 0;
+}
+
+template <typename I>
+int Migration<I>::v2_relink_src_image() {
+  ldout(m_cct, 10) << dendl;
+
+  int r = trash_restore(m_src_io_ctx, m_src_image_ctx->id, m_src_image_ctx->name);
+  if (r < 0) {
+    lderr(m_cct) << "failed restoring image from trash: " << cpp_strerror(r)
+                 << dendl;
+    return r;
+  }
+
+  return 0;
+}
+
+template <typename I>
+int Migration<I>::create_dst_image() {
+  ldout(m_cct, 10) << dendl;
+
+  uint64_t size;
+  {
+    RWLock::RLocker snap_locker(m_src_image_ctx->snap_lock);
+    size = m_src_image_ctx->size;
+  }
+
+  ThreadPool *thread_pool;
+  ContextWQ *op_work_queue;
+  ImageCtx::get_thread_pool_instance(m_cct, &thread_pool, &op_work_queue);
+
+  C_SaferCond on_create;
+  auto *req = image::CreateRequest<I>::create(
```

Comment:
@dillaman I believe you were reviewing a commit (not the final "file changes"), because it is updated by yours commit "librbd: migration should keep parent linkage by default"